Optimize the alignment format of inference results

zhixuwei opened the pull request:
Optimize the alignment format in which results.csv stores data, and improve the experience of viewing results directly in the text editor.
In practice, users often use the text editor that comes with the development tools to quickly see the results (results.csv), so alignment when storing data is necessary.


Fix [#16807](https://github.com/ultralytics/ultralytics/issues/16807)

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improves how training metrics are written to results.csv with auto-aligned, trimmed columns for cleaner, more readable logs. ✨

### 📊 Key Changes
- Added csv_col_width state to track column widths for results.csv formatting.
- Trimmed whitespace from CSV headers and values when reading results (ensures clean keys).
- Introduced save_format_csv to:
  - Auto-adjust column widths based on content.
  - Right-align and pad values for consistent spacing.
  - Rewrite the CSV with updated formatting when needed; otherwise append efficiently.
- Updated save_metrics to use the new formatted writing and standardized value formatting.

### 🎯 Purpose & Impact
- Better readability: results.csv now has neatly aligned columns, making it easier to scan metrics during and after training. 📈
- More robust parsing: removes stray whitespace in headers/values, reducing bugs in downstream analysis. 🧹
- Backward compatible: seamlessly formats existing files by trimming headers and reformatting as needed.
- Minimal overhead: only rewrites when column widths change; otherwise appends, keeping training fast. ⚡

zhixuwei committed:
Optimize the alignment format of inference results

Optimize the alignment format in which results.csv stores data, and improve the experience of viewing results directly in the text editor

UltralyticsAssistant commented:
👋 Hello @zhixuwei, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response. An Ultralytics engineer will also review your PR soon. Here's a quick checklist to help ensure a smooth integration:

- ✅ **Define a Purpose**: Your description clearly explains the purpose of enhancing CSV alignment and references [issue #16807](https://github.com/ultralytics/ultralytics/issues/16807). Great job! 👍
- ✅ **Synchronize with Source**: Ensure your PR is up-to-date with the `ultralytics/ultralytics` `main` branch. Merge any updates by clicking 'Update branch' or running `git pull` followed by `git merge main`.
- ✅ **Ensure CI Checks Pass**: Check that all Ultralytics [CI checks](https://docs.ultralytics.com/help/CI/) are successful. If there are issues, please resolve them. 
- ✅ **Update Documentation**: If your change impacts functionality, make sure to update the [documentation](https://docs.ultralytics.com) accordingly.
- ✅ **Add Tests**: Include tests that account for your changes, and ensure all existing tests pass.
- ✅ **Sign the CLA**: You've already indicated you signed the [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/)—thank you! 📝
- ✅ **Minimize Changes**: Keep alterations to the essential minimum required for clarity and function. Your approach aligns well with this principle, enhancing user experience without extraneous changes.

For more guidance, please refer to the [Contributing Guide](https://docs.ultralytics.com/help/contributing). If you have further questions, feel free to leave a comment. Thank you for enhancing Ultralytics! 🚀

zhixuwei commented:
Current format
![old](https://github.com/user-attachments/assets/87f94cf2-d5b1-4393-9508-39b6f7b5693d)

Optimize the format of the display
![new](https://github.com/user-attachments/assets/cc3769ba-dd72-431f-a43b-8af5066daca0)

Laughing-q commented:
@zhixuwei Thanks for the pr!
@glenn-jocher Actually I think you just removed the alignment we had in this PR: https://github.com/ultralytics/ultralytics/pull/16803. Is there a specific purpose for this? I feel the aligned results is more readable.

glenn-jocher commented:
@Laughing-q yes reason was that I created the aligned results a few years ago because I thought it was more readable but I've never seen any other CSVs like this, so I thought we might not be aligned with the CSV standard practices. Several users voiced complaints though, so it looks like I had it right originally.

Though I think this might require reverting a change in the checkpoint saving also, let me see.

glenn-jocher committed:
Merge branch 'main' into patch-3

glenn-jocher commented:
I guess the main issue is it's not a "real" CSV because of the spaces, i.e. the space format isn't read correctly out of the box by default tools:

```python
import pandas as pd

x = pd.read_csv("runs/detect/train7/results.csv").to_dict(orient="list")
print(x)

{'                  epoch': [1, 2, 3, 4, 5], '         train/box_loss': [1.7148, 2.1891, 1.2267, 1.7854, 1.7808], '         train/cls_loss': [4.9603, 5.4093, 5.2135, 5.7025, 5.7815], '         train/dfl_loss': [1.6274, 1.7632, 1.3243, 1.5428, 1.6947], '   metrics/precision(B)': [0.55624, 0.54948, 0.54057, 0.53095, 0.52424], '      metrics/recall(B)': [0.49102, 0.48393, 0.48402, 0.48495, 0.48412], '       metrics/mAP50(B)': [0.49929, 0.49917, 0.49816, 0.4987, 0.4987], '    metrics/mAP50-95(B)': [0.34924, 0.33244, 0.35222, 0.34811, 0.33389], '           val/box_loss': [1.7511, 1.777, 1.7996, 1.8286, 1.8355], '           val/cls_loss': [1.8236, 1.8448, 1.8505, 1.8666, 1.8796], '           val/dfl_loss': [1.2552, 1.2656, 1.274, 1.2851, 1.2842], '                 lr/pg0': [0.0, 9.5438e-07, 1.4375e-06, 1.4494e-06, 9.9008e-07], '                 lr/pg1': [0.0, 9.5438e-07, 1.4375e-06, 1.4494e-06, 9.9008e-07], '                 lr/pg2': [0.0, 9.5438e-07, 1.4375e-06, 1.4494e-06, 9.9008e-07]}
```

but with the recent PR that removed the spaces this is now read correctly:

```
import pandas as pd

x = pd.read_csv("runs/detect/train2/results.csv").to_dict(orient="list")
print(x)

{'epoch': [1, 2, 3], 'time': [0.134358, 0.468909, 2.77984], 'train/box_loss': [1.49337, 1.5282, 1.94261], 'train/cls_loss': [5.24029, 5.4021, 5.8874], 'train/dfl_loss': [1.56735, 1.57955, 1.68737], 'metrics/precision(B)': [0.0, 0.0, 0.00015], 'metrics/recall(B)': [0.0, 0.0, 0.01667], 'metrics/mAP50(B)': [0.0, 0.0, 0.00112], 'metrics/mAP50-95(B)': [0.0, 0.0, 0.00059], 'val/box_loss': [0.0, 0.0, 1.69917], 'val/cls_loss': [0.0, 0.0, 4.76302], 'val/dfl_loss': [0.0, 0.0, 1.30816], 'lr/pg0': [0.0, 7.973e-07, 8.092e-07], 'lr/pg1': [0.0, 7.973e-07, 8.092e-07], 'lr/pg2': [0.0, 7.973e-07, 8.092e-07]}
```

zhixuwei commented:
@glenn-jocher Your point is valid, as the current alignment is done by whitespace padding the original data, which means that the whitespace is part of the data and not an "adornment" that can be ignored during data processing. This also explains why there are Spaces in the results when processed by pandas.
However, in my usage, results.csv is a file that I use to quickly view the data related to the model training process. I have a habit of using the dev tools' text viewer or a separate text editor to take a quick look at the file. It's rare to use Excel or libraries like pandas or cvs. The main reason was that I wanted to see the results quickly and intuitively. Based on this situation, I think it is necessary to deal with the direct alignment of files.

Laughing-q committed:
update

Laughing-q commented:
@glenn-jocher @zhixuwei Yes actually I'm one of the users who check the results.csv sometimes so prefer it to be more readable. :)  I've reverted the model saving change as well in https://github.com/ultralytics/ultralytics/pull/16808/commits/b560c6c47b1da9fa2787d5c59729bf37ec6cbb16

zhixuwei commented:
> @glenn-jocher @zhixuwei Yes actually I'm one of the users who check the results.csv sometimes so prefer it to be more readable. :) I've reverted the model saving change as well in [b560c6c](https://github.com/ultralytics/ultralytics/commit/b560c6c47b1da9fa2787d5c59729bf37ec6cbb16)


We have the same idea.😊

glenn-jocher commented:
Thank you for your input! We'll keep your preference in mind as we continue to improve the user experience.

zhixuwei committed:
results.csv file readability optimization

Remove the fixed spacing between data columns and automatically adjust the spacing between data columns according to the change of data length. More parameters can be displayed while maintaining good readability.

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

zhixuwei commented:
previous
![image](https://github.com/user-attachments/assets/ed24aff3-c7f1-47cf-841e-5dbb1af7aa7b)

current
![image](https://github.com/user-attachments/assets/7bb01bf1-3bb4-40bd-b156-01fcf0f62105)

optimized
![image](https://github.com/user-attachments/assets/64c0a2d3-eea3-4806-a2dc-f01f04977a05)


@glenn-jocher @Laughing-q I have submitted a new update for CSV readability optimization. The fixed spacing between data columns is eliminated, and the strategy of automatically adjusting the spacing according to the actual width of data columns is adopted. This allows more columns to be displayed on the same screen, further improving readability.

glenn-jocher commented:
Thank you for the update! Your approach to optimizing CSV readability by adjusting spacing dynamically sounds promising. We'll review the changes and get back to you soon.

UltralyticsAssistant committed:
Merge branch 'main' into patch-3

UltralyticsAssistant committed:
Merge branch 'main' into patch-3

ambitious-octopus committed:
Merge branch 'main' into patch-3

UltralyticsAssistant committed:
Merge branch 'main' into patch-3

UltralyticsAssistant committed:
Merge branch 'main' into patch-3

UltralyticsAssistant committed:
Merge branch 'main' into patch-3

UltralyticsAssistant committed:
Merge branch 'main' into patch-3

zhixuwei committed:
Merge branch 'ultralytics:main' into patch-3

UltralyticsAssistant committed:
Merge branch 'main' into patch-3

UltralyticsAssistant committed:
Merge branch 'main' into patch-3

Y-T-G committed:
Merge branch 'main' into patch-3

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>

UltralyticsAssistant committed:
Merge branch 'main' into patch-3

UltralyticsAssistant committed:
Merge branch 'main' into patch-3
